File: sub_pub/flows/one_to_one.py
"""One-to-one flow: multiple source-destination pairs"""
import logging
import threading
from typing import List, Dict

from sub_pub.flows.base import Flow
from sub_pub.core.interfaces import MessageSource, MessagePublisher
from sub_pub.core.message import Message
from sub_pub.config.models import OneToOneMapping

logger = logging.getLogger(__name__)


class OneToOneFlow(Flow):
    """One-to-one flow: multiple independent source->destination mappings"""
    
    def __init__(
        self,
        source: MessageSource,
        destination: MessagePublisher,
        mappings: List[OneToOneMapping],
        **kwargs
    ):
        super().__init__(**kwargs)
        self.source = source
        self.destination = destination
        self.mappings = mappings
        
        # Build mapping dict for quick lookup
        self.topic_map: Dict[str, str] = {
            m.source_topic: m.destination_topic 
            for m in mappings
        }
        
    def _consume_messages(self) -> None:
        """Consume messages from source"""
        try:
            for message in self.source.consume():
                if not self._running:
                    break
                
                try:
                    # Check if this topic has a mapping
                    dest_topic = self.topic_map.get(message.topic)
                    if dest_topic is None:
                        logger.debug(f"No mapping found for topic {message.topic}, skipping")
                        continue
                    
                    # Record source metrics
                    self.metrics.record_source_message(message.topic, message.size())
                    
                    # Apply back-pressure if needed
                    if self._check_back_pressure(self.domain_queue):
                        logger.debug("Back-pressure detected on domain queue, waiting...")
                        self._wait_for_back_pressure_release(self.domain_queue)
                    
                    # Add to domain queue
                    self.domain_queue.put((message, dest_topic))
                    
                    # Commit the message
                    self.source.commit(message)
                    
                except Exception as e:
                    logger.error(f"Error processing message: {e}", exc_info=True)
                    self.metrics.record_source_error(message.topic)
        except Exception as e:
            logger.error(f"Error consuming messages: {e}", exc_info=True)
    
    def run(self) -> None:
        """Run the one-to-one flow"""
        self._running = True
        
        # Connect source and destination
        self.source.connect()
        source_topics = [m.source_topic for m in self.mappings]
        self.source.subscribe(source_topics)
        self.destination.connect()
        
        # Start domain workers
        domain_threads = []
        for i in range(self.thread_pool_config.max_workers):
            t = threading.Thread(
                target=self._domain_worker,
                name=f"domain-worker-{i}",
                daemon=True
            )
            t.start()
            domain_threads.append(t)
        
        # Start publish workers
        publish_threads = []
        for i in range(self.thread_pool_config.max_workers):
            t = threading.Thread(
                target=lambda: self._publish_worker(self.destination),
                name=f"publish-worker-{i}",
                daemon=True
            )
            t.start()
            publish_threads.append(t)
        
        # Start consumer thread
        consumer_thread = threading.Thread(
            target=self._consume_messages,
            name="consumer",
            daemon=True
        )
        consumer_thread.start()
        
        logger.info(f"One-to-one flow started with {len(self.mappings)} mappings")
        
        # Wait for shutdown signal
        self._shutdown_event.wait()
        
        # Cleanup
        consumer_thread.join(timeout=5.0)
        for thread in domain_threads + publish_threads:
            thread.join(timeout=5.0)
        
        # Flush and close
        self.destination.flush()
        self.destination.close()
        self.source.close()
        
        logger.info("One-to-one flow stopped")

File: rust/src/flows/one_to_one.rs
use std::collections::HashMap;
use std::sync::Arc;

use tracing::{error, info, warn};

use crate::config::OneToOneConfig;
use crate::domain::MessageProcessor;
use crate::error::SubPubError;
use crate::metrics::MetricsCollector;

use super::base::{apply_back_pressure, build_pipeline, spawn_domain_worker, PipelineItem};

/// One-to-one flow: each message arriving on a source topic is forwarded to
/// its configured destination topic, with optional processing in between.
///
/// Multiple topic-to-topic mappings can be served by a single flow instance —
/// messages whose source topic does not appear in `mappings` are silently
/// skipped.
pub struct OneToOneFlow {
    cfg: OneToOneConfig,
    processor: Arc<dyn MessageProcessor>,
    metrics: Arc<MetricsCollector>,
}

impl OneToOneFlow {
    pub fn new(
        cfg: OneToOneConfig,
        processor: Arc<dyn MessageProcessor>,
        metrics: Arc<MetricsCollector>,
    ) -> Self {
        Self {
            cfg,
            processor,
            metrics,
        }
    }

    /// Run the flow until `shutdown_rx` fires or a fatal error occurs.
    pub async fn run(
        self,
        shutdown_rx: tokio::sync::watch::Receiver<bool>,
    ) -> Result<(), SubPubError> {
        use crate::adapters::{create_publisher, create_source};

        // Build the topic-to-topic mapping for fast lookup.
        let mapping: HashMap<String, String> = self
            .cfg
            .mappings
            .iter()
            .map(|m| (m.source_topic.clone(), m.destination_topic.clone()))
            .collect();

        if mapping.is_empty() {
            return Err(SubPubError::Config(
                "OneToOneConfig: no topic mappings defined".to_string(),
            ));
        }

        // Collect all source topics.
        let source_topics: Vec<String> = mapping.keys().cloned().collect();
        let source_topic_refs: Vec<&str> = source_topics.iter().map(String::as_str).collect();

        // Build source and publisher.
        let mut source = create_source(&self.cfg.source)?;
        let mut publisher = create_publisher(&self.cfg.destination)?;
        source.connect().await?;
        source.subscribe(&source_topic_refs).await?;
        publisher.connect().await?;

        let queue_size = self.cfg.thread_pool.queue_size;
        let (domain_tx, publish_tx, domain_rx, mut publish_rx) =
            build_pipeline(queue_size, queue_size);

        // Spawn domain-processing worker.
        let _domain_handle = spawn_domain_worker(
            domain_rx,
            publish_tx.clone(),
            Arc::clone(&self.processor),
            Arc::clone(&self.metrics),
        );

        // Spawn publish worker.
        let metrics_pub = Arc::clone(&self.metrics);
        let _publish_handle = tokio::spawn(async move {
            while let Some(item) = publish_rx.recv().await {
                let topic = &item.destination_topic;
                let size = item.message.size();
                match publisher.publish(&item.message, topic).await {
                    Ok(()) => metrics_pub.record_destination_message(topic, size),
                    Err(e) => {
                        error!(%e, %topic, "publish error");
                        metrics_pub.record_destination_error(topic);
                    }
                }
            }
            let _ = publisher.flush().await;
            let _ = publisher.close().await;
        });

        info!("OneToOneFlow started; topics: {:?}", source_topics);

        // Consumer loop.
        loop {
            // Honour shutdown signal.
            if *shutdown_rx.borrow() {
                break;
            }

            apply_back_pressure(&self.cfg.back_pressure, &domain_tx).await;

            match source.consume().await {
                Ok(Some(msg)) => {
                    let src_topic = msg.topic.clone();
                    match mapping.get(&src_topic) {
                        Some(dest_topic) => {
                            let size = msg.size();
                            self.metrics.record_source_message(&src_topic, size);
                            let dest = dest_topic.clone();
                            if domain_tx
                                .send(PipelineItem {
                                    message: msg,
                                    destination_topic: dest,
                                })
                                .await
                                .is_err()
                            {
                                warn!("domain channel closed; stopping consumer");
                                break;
                            }
                        }
                        None => {
                            warn!(%src_topic, "no mapping for topic; skipping");
                        }
                    }
                }
                Ok(None) => {
                    info!("source exhausted; stopping flow");
                    break;
                }
                Err(e) => {
                    error!(%e, "consume error");
                    self.metrics.record_source_error("unknown");
                }
            }

            // Check shutdown again after potentially blocking on consume.
            if shutdown_rx.has_changed().unwrap_or(false) && *shutdown_rx.borrow() {
                break;
            }
        }

        // Graceful shutdown: close the domain channel so downstream workers
        // drain their queues before exiting.
        drop(domain_tx);
        let _ = source.close().await;
        info!("OneToOneFlow stopped");
        Ok(())
    }
}
